tools/please_pex/preamble/pexerror.c

```diff
@@ -0,0 +1,98 @@
+/*
+ * pexerror.c implements a simple error reporting mechanism.
+ *
+ * Errors are singly-linked lists (of type err_t) whose nodes (of type errmsg_t) are strings
+ * containing diagnostic messages. This allows for the formation of a Go-like error chain, with
+ * diagnostic messages becoming progressively more specific as the tail of the list is approached.
+ */
+
+#include <errno.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sysexits.h>
+
+#include "pexerror.h"
+#include "util.h"
+
+/*
+ * err_from_str returns an error from the given diagnostic message.
+ */
+err_t *err_from_str(const char *msg) {
+    err_t *err = NULL;
+
+    MALLOC(err, err_t, 1);
+    *err = (err_t){ 0 };
+    SLIST_INIT(err);
+
+    return err_wrap(msg, err);
+}
+
+/*
+ * err_from_errno returns an error from the given diagnostic message that wraps another error
+ * whose diagnotic message is derived from the current value of errno.
+ *
+ * This function is used to report the failure of a system call or library function. Because errno
+ * doesn't indicate which system call or library function encountered the error, the diagnostic
+ * message passed in msg typically includes the name of the system call or library function.
+ *
+ * Note that err_from_errno doesn't check that the value of errno is significant; it is therefore
+ * the responsibility of the caller to ensure that a system call or library function that sets
+ * errno has actually returned a value indicating failure before calling err_from_errno.
+ */
+err_t *err_from_errno(const char *msg) {
+    err_t *err = err_from_str(strerror(errno));
+
+    return err_wrap(msg, err);
+}
+
+/*
+ * err_wrap creates an error from the given diagnostic message, using it to wrap the given error,
+ * and then returns the newly-created error.
+ */
+err_t *err_wrap(const char *msg, err_t *wrapped) {
+    errmsg_t *first = SLIST_FIRST(wrapped);
+    errmsg_t *errmsg = NULL;
+    size_t len = strlen(msg);
+
+    if (wrapped == NULL) {
```

Comment:
should this check be before line 53? seems odd to check for nullness when you've already used `wrapped`